river/base/adpredictor.py

```diff
@@ -0,0 +1,154 @@
+from __future__ import annotations
+
+from collections import defaultdict, namedtuple
```

Comment:
We prefer to import the package, and then access its properties, instead of importing what we need

```suggestion
import collections
```